lib/puppet/settings/file_setting.rb

```diff
@@ -155,10 +155,24 @@ def to_resource
         resource[:mode] = mode
       end
 
-      # REMIND fails on Windows because chown/chgrp functionality not supported yet
-      if Puppet.features.root? and !Puppet.features.microsoft_windows?
-        resource[:owner] = self.owner if self.owner
-        resource[:group] = self.group if self.group
+      if Puppet.features.root?
+        if Puppet.features.microsoft_windows?
+          # When running as an Administrator it is possible to create settings files and directories
+          # that are only owned by a single Administrator.  This may make sense for normal file
+          # resources, it doesn't really apply to Puppet settings files. Therefore if we are running
+          # with elevated rights (Puppet.features.root?) and on Windows (Puppet.features.microsoft_windows?)
+          # then all settings files and directories should be owned by the Administrators group, with the primary
+          # group of LOCAL SYSTEM.  This is the typical configuration for files created as service logged on as
+          # LOCAL SYSTEM. Note that the LOCAL SYSTEM user is always an implicit member of the Administrators group
+          # therefore they don't needed to be added to the resultant ACL.  Note we also use the SID representation
+          # to avoid localization issues or account renaming.
+          resource[:owner] = 'S-1-5-32-544'
+          resource[:group] = 'S-1-5-18'
```

Comment:
So I don't think that this is quite the right spot to do this, given how the `owner` and `group` values are populated / set in the first place by `Root.new` or `Service.new`. Instead, I think we want to augment the `Root.new` instance to return something other than 'root' correctly based on OS, rather than ignoring the values for `owner` and `group` that have been set on this instance. That allows us to remove this guard completely (similar to what was done in this ancient PR - https://github.com/puppetlabs/puppet/pull/2153).

I'm also thinking that we can leave the default `group` set to `nil` to get the behavior that we want. That should translate to `Administrators: (F)` and `SYSTEM: (F)` being set by the security code. If we were to set an explicit `group`, it probably makes sense to instead use something like the `Service` well-known SID `S-1-5-6` (like that old PR), so that any service is set for the `group` and not just `SYSTEM`.

I'm hoping an explicit group is not necessary, but one thing to test for sure here is the use of a domain service account (is it possible to not make it part of local `Administrators`?).

Reply:
FWIW - I see what you're describing for creating a dir as `SYSTEM` under `ProgramData` like:

```
Owner                   : BUILTIN\Administrators
Group                   : NT AUTHORITY\SYSTEM
```

But I haven't yet tested a virtual account or a domain account, etc.

Reply:
Originally I didn't manage `group` but it is possible then to have the `owner` and `group` be the same, which will then cause log/IO churn because some of the settings are mode `750`.

> a domain service account (is it possible to not make it part of local Administrators?)
Yes, it is possible to do this. However this is fine.  It means that `Puppet.features.root?` will be false and will instead use `%USERPROFILE%\.puppetlabs` for the settings not Programdata.

Same with Virtual Accounts, although off-hand I don't know what %USERPROFILE% is for a virtual account.

Reply:
> However this is fine. It means that `Puppet.features.root?` will be false and will instead use `%USERPROFILE%\.puppetlabs` for the settings not `Programdata`.

It's not quite that simple. `root?` / `admin?` is not determined by group membership, but rather by token. See the `elevated_security?` implementation which checks for a specific elevated flag in the token at https://github.com/puppetlabs/puppet/blob/3a5dfed0fef8fc5cb1b6ec168766115de3e8f23b/lib/puppet/util/windows/process.rb#L183-L205

You can use `psexec` to log in as `Local Service` or `Network Service`, but both of those accounts are already members of `BUILTIN\Administrators`, which would already be granted `(F)`. Virtual accounts cannot be logged in with, so I'm not sure if they're actually members of `Administrators` or not.

I was able to create a user that is not a member of `Administrators` group, but does have an elevated token. Since Puppet considers that user `root?`, it wrote under `ProgramData`. It was unable to read / write a file that had been setup with permissions of `Administrators:(F)` / `SYSTEM:(F)`.

Running `psexec` as `SYSTEM`, I was able to create files / dirs on disk with the perms setup the way the PR intends:

```
PS C:\ProgramData> get-acl .\Foobar\ | select *


PSPath                  : Microsoft.PowerShell.Core\FileSystem::C:\ProgramData\Foobar\
PSParentPath            : Microsoft.PowerShell.Core\FileSystem::C:\ProgramData
PSChildName             : Foobar
PSDrive                 : C
PSProvider              : Microsoft.PowerShell.Core\FileSystem
CentralAccessPolicyId   :
CentralAccessPolicyName :
Path                    : Microsoft.PowerShell.Core\FileSystem::C:\ProgramData\Foobar\
Owner                   : BUILTIN\Administrators
Group                   : NT AUTHORITY\SYSTEM
Access                  : {System.Security.AccessControl.FileSystemAccessRule,
                          System.Security.AccessControl.FileSystemAccessRule}
Sddl                    : O:BAG:SYD:PAI(A;OICI;FA;;;SY)(A;OICI;FA;;;BA)
AccessToString          : NT AUTHORITY\SYSTEM Allow  FullControl
                          BUILTIN\Administrators Allow  FullControl
AuditToString           :
AccessRightType         : System.Security.AccessControl.FileSystemRights
AccessRuleType          : System.Security.AccessControl.FileSystemAccessRule
AuditRuleType           : System.Security.AccessControl.FileSystemAuditRule
AreAccessRulesProtected : True
AreAuditRulesProtected  : False
AreAccessRulesCanonical : True
AreAuditRulesCanonical  : True

PS C:\ProgramData> get-acl .\Foobar\test.txt | select *

PSPath                  : Microsoft.PowerShell.Core\FileSystem::C:\ProgramData\Foobar\test.txt
PSParentPath            : Microsoft.PowerShell.Core\FileSystem::C:\ProgramData\Foobar
PSChildName             : test.txt
PSDrive                 : C
PSProvider              : Microsoft.PowerShell.Core\FileSystem
CentralAccessPolicyId   :
CentralAccessPolicyName :
Path                    : Microsoft.PowerShell.Core\FileSystem::C:\ProgramData\Foobar\test.txt
Owner                   : BUILTIN\Administrators
Group                   : NT AUTHORITY\SYSTEM
Access                  : {System.Security.AccessControl.FileSystemAccessRule,
                          System.Security.AccessControl.FileSystemAccessRule}
Sddl                    : O:BAG:SYD:AI(A;ID;FA;;;BA)(A;ID;FA;;;SY)
AccessToString          : BUILTIN\Administrators Allow  FullControl
                          NT AUTHORITY\SYSTEM Allow  FullControl
AuditToString           :
AccessRightType         : System.Security.AccessControl.FileSystemRights
AccessRuleType          : System.Security.AccessControl.FileSystemAccessRule
AuditRuleType           : System.Security.AccessControl.FileSystemAuditRule
AreAccessRulesProtected : False
AreAuditRulesProtected  : False
AreAccessRulesCanonical : True
AreAuditRulesCanonical  : True
```


```
Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

C:\Windows\system32>whoami /all

USER INFORMATION
----------------

User Name                    SID
============================ ============================================
vagrant-2008r2\adminbytokens S-1-5-21-271343509-1886877197-423808128-4888


GROUP INFORMATION
-----------------

Group Name                                           Type             SID                                   Attributes

==================================================== ================ 
========================================================================================
=====
Everyone                                             Well-known group S-1-1-0
                                   Mandatory group, Enabled by default, Enabled

VAGRANT-2008R2\g45991a14-ee2d-48f6-925c-6ea809a5f994 Alias            S-1-5-21-271343509-1886877197-423808128-4735 
                                   Mandatory group, Enabled by default, Enabled

VAGRANT-2008R2\TestGroup-PUP8231                     Alias            S-1-5-21-271343509-1886877197-423808128-4699 Mandatory group, Enabled by default, Enabled

BUILTIN\Users                                        Alias            S-1-5-32-545                                 Mandatory group, Enabled by default, Enabled

NT AUTHORITY\INTERACTIVE                             Well-known group S-1-5-4                                   Mandatory group, Enabled by default, Enabled

CONSOLE LOGON                                        Well-known group S-1-2-1                                   Mandatory group, Enabled by default, Enabled

NT AUTHORITY\Authenticated Users                     Well-known group S-1-5-11                                   Mandatory group, Enabled by default, Enabled

NT AUTHORITY\This Organization                       Well-known group S-1-5-15                                   Mandatory group, Enabled by default, Enabled

NT AUTHORITY\Local account                           Well-known group S-1-5-113                                   Mandatory group, Enabled by default, Enabled

NT AUTHORITY\NTLM Authentication                     Well-known group S-1-5-64-10                                  Mandatory group, Enabled by default, Enabled

Mandatory Label\High Mandatory Level                 Label            S-1-16-12288                                 Mandatory group, Enabled by default, Enabled



PRIVILEGES INFORMATION
----------------------

Privilege Name                  Description                               State

=============================== ========================================= ========
SeIncreaseQuotaPrivilege        Adjust memory quotas for a process        Disabled
SeSecurityPrivilege             Manage auditing and security log          Disabled
SeTakeOwnershipPrivilege        Take ownership of files or other objects  Disabled
SeLoadDriverPrivilege           Load and unload device drivers            Disabled
SeSystemProfilePrivilege        Profile system performance                Disabled
SeSystemtimePrivilege           Change the system time                    Disabled
SeProfileSingleProcessPrivilege Profile single process                    Disabled
SeIncreaseBasePriorityPrivilege Increase scheduling priority              Disabled
SeCreatePagefilePrivilege       Create a pagefile                         Disabled
SeBackupPrivilege               Back up files and directories             Disabled
SeRestorePrivilege              Restore files and directories             Disabled
SeShutdownPrivilege             Shut down the system                      Disabled
SeDebugPrivilege                Debug programs                            Disabled
SeSystemEnvironmentPrivilege    Modify firmware environment values        Disabled
SeChangeNotifyPrivilege         Bypass traverse checking                  Enabled
SeRemoteShutdownPrivilege       Force shutdown from a remote system       Disabled
SeUndockPrivilege               Remove computer from docking station      Disabled
SeManageVolumePrivilege         Perform volume maintenance tasks          Disabled
SeImpersonatePrivilege          Impersonate a client after authentication Enabled
SeCreateGlobalPrivilege         Create global objects                     Enabled
SeIncreaseWorkingSetPrivilege   Increase a process working set            Disabled
SeTimeZonePrivilege             Change the time zone                      Disabled
SeCreateSymbolicLinkPrivilege   Create symbolic links                     Disabled

C:\Windows\system32>cd \source\PUP-6729-testing

C:\source\PUP-6729-testing>cd ..\puppet

[1] pry(main)> require 'puppet'
=> true
[2] pry(main)> Puppet::Util::Windows::Process.elevated_security?
=> true
[3] pry(main)> exit

C:\source\puppet>cd \ProgramData

C:\ProgramData>dir Foobar
 Volume in drive C is Windows 2008
 Volume Serial Number is D8AA-E11A

 Directory of C:\ProgramData\Foobar

File Not Found

C:\ProgramData>icacls Foobar
Foobar: Access is denied.
Successfully processed 0 files; Failed processing 1 files

C:\ProgramData>type Foobar\test.txt
Access is denied.
```

We've also been able to repro another broken scenario by running puppet as this non-admin user and getting it to write files to `\ProgramData\PuppetLabs`, but may have to file a separate issue on that one as I'm not certain it's even worth chasing down. (i.e. when this is done it prevents a local Administrator from accessing `client_data`)

If we actively manage the `group` with Puppet, I'm still leaning toward letting Windows default it, because then there's a better chance that a weird scenario will allow the word scenario to continue working, in addition to still granting `Administrators:(F)` *and* `SYSTEM:(F)` correctly. Since we know we always drop that `SYSTEM:(F)`, I think we can rely on that behavior instead of setting the group.